gnocchi/rest/api.py

```diff
@@ -1890,6 +1899,102 @@ class BatchController(object):
+        (indexer.NoSuchResource, indexer.ResourceAlreadyExists,
+         indexer.NamedMetricAlreadyExists)))
+    def get_or_create_resource_and_metrics(cls, creator, rid,
+                                           original_resource_id,
+                                           job, instance, metric_names):
+        try:
+            r = pecan.request.indexer.get_resource('prometheus', rid,
+                                                   with_metrics=True)
+        except indexer.NoSuchResourceType:
+            schema = pecan.request.indexer.get_resource_type_schema()
+            rt = schema.resource_type_from_dict(
+                'prometheus', cls.PROMETHEUS_RESOURCE_TYPE, 'creating')
+            pecan.request.indexer.create_resource_type(rt)
+            raise tenacity.TryAgain
+        except indexer.UnexpectedResourceTypeState as e:
+            # NOTE(sileht): Currently created by another thread
+            if not e.state.endswith("_error"):
+                raise tenacity.TryAgain
+
+        if r:
+            exists_metric_names = [m.name for m in r.metrics]
+            metrics = MetricsSchema(dict(
+                (m, {}) for m in metric_names
+                if m not in exists_metric_names
+            ))
+            if metrics:
+                return pecan.request.indexer.update_resource(
+                    'prometheus', rid,
+                    metrics=metrics,
+                    append_metrics=True,
+                    create_revision=False
+                ).metrics
+            else:
+                return r.metrics
+        else:
+            metrics = MetricsSchema(dict((m, {}) for m in metric_names))
+            return pecan.request.indexer.create_resource(
+                'prometheus', rid, creator,
+                original_resource_id=original_resource_id,
+                job=job,
+                instance=instance,
+                metrics=metrics
+            ).metrics
+
+    @pecan.expose()
+    def post(self):
+        buf = snappy.uncompress(pecan.request.body)
+        f = remote_pb2.WriteRequest()
+        f.ParseFromString(buf)
+        measures_by_rid = collections.defaultdict(dict)
+        for ts in f.timeseries:
+            attrs = dict((l.name, l.value) for l in ts.labels)
+            original_rid = (attrs["job"], attrs["instance"])
+            if ts.samples:
+                measures_by_rid[original_rid][attrs['__name__']] = (
+                    MeasuresListSchema([{'timestamp': s.timestamp_ms / 1000.0,
+                                         'value': s.value}
+                                        for s in ts.samples]))
+
+        creator = pecan.request.auth_helper.get_current_user(pecan.request)
+
+        measures_to_batch = {}
+        for (job, instance), measures in measures_by_rid.items():
+            original_rid = '%s@%s' % (job, instance)
+            rid = ResourceUUID(original_rid, creator=creator)
+            metric_names = list(measures.keys())
```

Comment:
It's possible Prometheus supports `/` in metric name and Gnocchi does not. You need to do a replace I guess and document it.

Reply:
I missed that, fixed.

Reply:
I missed that, fixed.